WikiWatershed/model-my-watershed
# PA DEP Celery Workflow

## Overview

Adds a Celery workflow that

 - Takes an area of interest (AoI)
 - Matches it to one or more HUC-12s
 - In case of more than one HUC-12, clips the AoI to each HUC-12 and pairs it with the HUC-12
 - For each AoI, runs NLCD and Streams analyses
 - For each HUC-12, runs NLCD and Streams analyses, and also GWLF-E to get land category specific water quality loads
 - Assembles all these values into a JSON object
 - This object can then be POSTed to `/export/worksheet/` to get a ZIP file containing an Excel file per HUC-12, and the shape of the clipped AoI as geojson

Connects #3192 
Connects #3193 

### Demo

![image](https://user-images.githubusercontent.com/1430060/70359940-1f011580-184b-11ea-86f0-e7a10272eda8.png)

### Notes

~There are certain odd cases where I'm seeing the return of the dreaded `chord_unlock` cycle. I will make a separate card to investigate that.~

We avoid this entirely by changing the Celery workflow to be linear rather than chorded in 09731c5.

## Testing Instructions

* Check out this branch and reprovision `app` and `worker`
* Run `debugcelery`
* Login to MMW and find your API Key (under My Account → Account)
* Save it as an environment variable
* Download this shape: [northwales.geojson.txt](https://github.com/WikiWatershed/model-my-watershed/files/3934111/northwales.geojson.txt)
* Use that shape to POST to `/api/modeling/worksheet/` with your API Token:

```http
http :8000/api/modeling/worksheet/ Authorization:"Token $MMW_API_TOKEN" < northwales.geojson

HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Encoding: gzip
Content-Type: application/json
Date: Fri, 06 Dec 2019 20:05:02 GMT
Location: /api/jobs/b7986887-c2f4-4cd8-959f-c1167ed00af1/
Server: nginx
Transfer-Encoding: chunked
Vary: Accept-Encoding
Vary: Accept, Cookie, Origin

{
    "job": "b7986887-c2f4-4cd8-959f-c1167ed00af1",
    "status": "started"
}
```

* Save the results of that job to a file

```console
http :8000/api/jobs/b7986887-c2f4-4cd8-959f-c1167ed00af1/ Authorization:"Token $MMW_API_TOKEN" | jq .result > northwales-payload.json
```

* Send that payload to the export endpoint and save the result:

```http
http --download :8000/export/worksheet/ < northwales-payload.json

HTTP/1.1 200 OK
Allow: POST, OPTIONS
Connection: keep-alive
Content-Disposition: attachment; filename="MMW_BMP_Spreadsheets.zip"
Content-Length: 82889
Content-Type: application/zip
Date: Fri, 06 Dec 2019 21:37:28 GMT
Server: nginx
Vary: Accept, Cookie

Downloading 80.95 kB to "MMW_BMP_Spreadsheets.zip"
Done. 80.95 kB in 0.00205s (38.63 MB/s)
```

* Ensure the file contents are identical to this: [MMW_BMP_Spreadsheets.zip](https://github.com/WikiWatershed/model-my-watershed/files/3934205/MMW_BMP_Spreadsheets.zip)
* Ensure the square meter block is populated, as should be the case for any AoI less than 2 square kilometer in size
* Try another shape, perhaps one that intersects multiple HUC-12s. Ensure the results are expected.
